### ISSUE TITLE
Treeview - large dataset rendering issue

### ISSUE THREAD

**Issue by Nodios** [telerik/kendo-react#334]
## I'm submitting a...

* Bug report

## Current behavior

We are using treeview to display a dataset with 3 levels of nesting (>5000 items). Filtering is mandatory thus we are filtering trough entire dataset to find relevant items. Filtered items are also expanded only if they have children. This creates a lag in treeview render (we are looking at about aprox. 2-3 seconds of nothing)

## Minimal reproduction of the problem with instructions
Here is a link. This is repro based on your example, it is not 3 level nesting, but it will serve a purpose.
https://stackblitz.com/edit/react-xz7cjb

**Comment by simonssspirit** [comment]
In cases like this one, we recommend loading on demand. Still, if there 5000 items, the speed will depend on the machine and the browser, as well as this, is heavy operation. 

Also, React is times faster when build for production. If the same example is built for production, the expanding of all items will be noticeably faster.

**Comment by Nodios** [comment]
Loading on demand is a way to go definitely, but is not applicable in my case since I have to search through the entire dataset. 

Search without expanding found items works in less than a second. But if I want to expand found items, search is done in ignorable time, but the rendering takes a noticeable amount of time.

**Comment by simonssspirit** [comment]
Is server filtering possible? For example, send the filter to the server, and return only the filtered nodes with the parent items already expanded. This should make the tree smaller and faster to render.

**Comment by Nodios** [comment]
Server filtering was the first thing we considered. Thing is that this is (only) one part of a system that is under heavy usage. We have multiple parts of the system where we're also going to have treeview component under heavy usage. This would put too much strain on the server. 

I missed your production comment, can you elaborate on it? Aren't your components already built for production?

**Comment by simonssspirit** [comment]
Sorry if my statement was not clear.  Our components are built for production. The React application itself is faster when build for production(e.g npm run build). The compiled code that will be used in production is faster than the one in development and StackBlitz.

**Comment by Nodios** [comment]
That is true, yeah. I did the production build, and it is noticeably faster! Thanks for the hints
